### PR TITLE
[Merged by Bors] - feat: Star algebras satisfy `OrderedSMul`

### PR DESCRIPTION
**Pull request by dupuisf**
This PR shows that star algebras satisfy `OrderedSMul 𝕜 A` whenever `𝕜` is `RCLike` and `A` is a non-unital star algebra.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 1]
```diff
@@ -881,14 +881,35 @@ lemma ofReal_le_ofReal {x y : ℝ} : (x : K) ≤ (y : K) ↔ x ≤ y := by
   rw [le_iff_re_im]
   simp
 
+@[simp]
```
Comment (urkud) [inline Mathlib/Analysis/RCLike/Basic.lean:884]: Should it me `norm_cast` too?

Comment (dupuisf) [reply Mathlib/Analysis/RCLike/Basic.lean]: Yes, seems like a good idea -- done!

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 2]
```diff
@@ -902,6 +902,18 @@ lemma ofReal_pos {x : ℝ} : 0 < (x : K) ↔ 0 < x := by
 lemma ofReal_lt_zero {x : ℝ} : (x : K) < 0 ↔ x < 0 := by
   rw [← ofReal_zero, ofReal_lt_ofReal]
 
+lemma inv_pos_of_pos (hz : 0 < z) : 0 < z⁻¹ := by
+  rw [pos_iff_exists_ofReal] at hz
+  obtain ⟨x, hx, hx'⟩ := hz
+  rw [← hx', ← ofReal_inv, ofReal_pos]
+  exact _root_.inv_pos_of_pos hx
+
+lemma inv_pos : 0 < z⁻¹ ↔ 0 < z := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/RCLike/Basic.lean:911]: Probably this and `inv_pos_of_pos` should be protected? That way we can avoid the `_root_.inv_pos` bits. If you really think it won't come up that often, you can leave it and mark this as resolved.

File: Mathlib/Algebra/Star/Order.lean [hunk 3]
```diff
@@ -256,6 +256,56 @@ lemma star_lt_one_iff {x : R} : star x < 1 ↔ x < 1 := by
...
+    refine AddSubmonoid.closure_induction hab ?mem ?zero ?add
+    case mem =>
+      intro x hx
+      have hc := le_of_lt hc
+      rw [StarOrderedRing.nonneg_iff] at hc
+      refine AddSubmonoid.closure_induction hc ?memc ?zeroc ?addc
+      case memc =>
+        intro c' hc'
+        rw [Set.mem_range] at hc'
+        rw [Set.mem_range] at hx
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/Order.lean:282]:
```suggestion
```

File: Mathlib/Algebra/Star/Order.lean [hunk 4]
```diff
@@ -256,6 +256,56 @@ lemma star_lt_one_iff {x : R} : star x < 1 ↔ x < 1 := by
...
+        intro c' d hc' hd
+        simp only [add_smul]
+        exact AddSubmonoid.add_mem (AddSubmonoid.closure (Set.range fun s => star s * s)) hc' hd
+    case zero =>
+      simp only [smul_zero]
+      exact AddSubmonoid.zero_mem (AddSubmonoid.closure (Set.range fun s => star s * s))
+    case add =>
+      intro x y hx hy
+      rw [smul_add]
+      exact AddSubmonoid.add_mem (AddSubmonoid.closure (Set.range fun s => star s * s)) hx hy
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/Order.lean:301]: Make your life easy, use the generic versions and let Lean infer the arguments:
```
      case zeroc => simpa only [zero_smul] using zero_mem _
      case addc => exact fun c' d ↦ by simpa only [add_smul] using add_mem
    case zero => simpa only [smul_zero] using zero_mem _
    case add => exact fun x y ↦ by simpa only [smul_add] using add_mem
```

File: Mathlib/Analysis/RCLike/Basic.lean (hunk 2 above)
Comment (dupuisf) [reply Mathlib/Analysis/RCLike/Basic.lean:911]: Fair point; I've made them protected.

File: Mathlib/Analysis/InnerProductSpace/LaxMilgram.lean [hunk 5]
```diff
@@ -64,10 +64,10 @@ theorem bounded_below (coercive : IsCoercive B) : ∃ C, 0 < C ∧ ∀ v, C * 
 
 theorem antilipschitz (coercive : IsCoercive B) : ∃ C : ℝ≥0, 0 < C ∧ AntilipschitzWith C B♯ := by
   rcases coercive.bounded_below with ⟨C, C_pos, below_bound⟩
-  refine ⟨C⁻¹.toNNReal, Real.toNNReal_pos.mpr (inv_pos.mpr C_pos), ?_⟩
+  refine ⟨C⁻¹.toNNReal, Real.toNNReal_pos.mpr (_root_.inv_pos.mpr C_pos), ?_⟩
```
Comment (j-loreaux) [inline Mathlib/Analysis/InnerProductSpace/LaxMilgram.lean:67]: This is no longer necessary, right? 
```suggestion
  refine ⟨C⁻¹.toNNReal, Real.toNNReal_pos.mpr (inv_pos.mpr C_pos), ?_⟩
```

File: Mathlib/Analysis/InnerProductSpace/LaxMilgram.lean [hunk 6]
```diff
@@ -64,10 +64,10 @@ theorem bounded_below (coercive : IsCoercive B) : ∃ C, 0 < C ∧ ∀ v, C * 
...
   refine ContinuousLinearMap.antilipschitz_of_bound B♯ ?_
-  simp_rw [Real.coe_toNNReal', max_eq_left_of_lt (inv_pos.mpr C_pos), ←
-    inv_mul_le_iff (inv_pos.mpr C_pos)]
+  simp_rw [Real.coe_toNNReal', max_eq_left_of_lt (_root_.inv_pos.mpr C_pos), ←
+    inv_mul_le_iff (_root_.inv_pos.mpr C_pos)]
```
Comment (j-loreaux) [inline Mathlib/Analysis/InnerProductSpace/LaxMilgram.lean:70]:
```suggestion
  simp_rw [Real.coe_toNNReal', max_eq_left_of_lt (inv_pos.mpr C_pos), ←
    inv_mul_le_iff (inv_pos.mpr C_pos)]
```

File: Mathlib/Analysis/NormedSpace/PiLp.lean [hunk 7]
```diff
@@ -200,7 +200,8 @@ protected theorem edist_self (f : PiLp p β) : edist f f = 0 := by
   rcases p.trichotomy with (rfl | rfl | h)
   · simp [edist_eq_card]
   · simp [edist_eq_iSup]
-  · simp [edist_eq_sum h, ENNReal.zero_rpow_of_pos h, ENNReal.zero_rpow_of_pos (inv_pos.2 <| h)]
+  · simp [edist_eq_sum h, ENNReal.zero_rpow_of_pos h, ENNReal.zero_rpow_of_pos
+          (_root_.inv_pos.2 <| h)]
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/PiLp.lean:204]:
```suggestion
  · simp [edist_eq_sum h, ENNReal.zero_rpow_of_pos h, ENNReal.zero_rpow_of_pos (inv_pos.2 <| h)]
```

File: Mathlib/Analysis/NormedSpace/ProdLp.lean [hunk 8]
```diff
@@ -190,7 +190,7 @@ theorem prod_edist_self (f : WithLp p (α × β)) : edist f f = 0 := by
     simp
   · simp [prod_edist_eq_sup]
   · simp [prod_edist_eq_add h, ENNReal.zero_rpow_of_pos h,
-      ENNReal.zero_rpow_of_pos (inv_pos.2 <| h)]
+      ENNReal.zero_rpow_of_pos (_root_.inv_pos.2 <| h)]
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/ProdLp.lean:193]:
```suggestion
      ENNReal.zero_rpow_of_pos (inv_pos.2 <| h)]
```

**Review by j-loreaux — COMMENTED** [review]
I think we can remove`_root_.` now. 

bors d+

**Pull request closed without merging**